src/cpp/session/modules/shiny/ShinyAsyncJob.cpp

```diff
@@ -0,0 +1,133 @@
+ *
+ */
+
+#define kShinyAppStarted "Shiny started; listening on URL: " 
+
+#include "ShinyAsyncJob.hpp"
+#include "../SessionShinyViewer.hpp"
+
+#include <session/jobs/JobsApi.hpp>
+
+#include <session/SessionUrlPorts.hpp>
+#include <session/SessionModuleContext.hpp>
+
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules { 
+namespace shiny {
+
+ShinyAsyncJob::ShinyAsyncJob(const std::string& name, 
+      const FilePath& path,
+      const std::string& viewerType,
+      const std::string& runCmd):
+   AsyncRJob(name),
+   path_(path),
+   viewerType_(viewerType),
+   runCmd_(runCmd)
+{
+}
+
+void ShinyAsyncJob::start()
+{
+   // set a status before we start R
+   jobs::setJobStatus(job_, "Starting");
+
+   // create the string to send to R
+   std::string cmd(
+         "options(shiny.launch.browser = function(url) { "
+         "   cat(\"" kShinyAppStarted "\", url)"
+         "}); " +
+         runCmd_);
+
+   // start the R process
+   core::system::Options environment;
+   async_r::AsyncRProcess::start(cmd.c_str(), environment, path_.getParent(), 
+         async_r::AsyncRProcessOptions::R_PROCESS_NO_RDATA);
+
+   // echo the command we submitted to R
+   onStdout("=> " + runCmd_ + "\n\n");
+}
+
+void ShinyAsyncJob::enqueueStateEvent(const std::string& state)
+{
+   json::Object dataJson;
+   dataJson["url"] = url_ports::mapUrlPorts(url_);
+   dataJson["path"] = module_context::createAliasedPath(path_);
+   dataJson["state"] = state;
+   dataJson["viewer"] = viewerType_;
+   dataJson["options"] = shiny_viewer::SHINY_VIEWER_OPTIONS_NONE;
+   dataJson["id"] = job_->id();
+   ClientEvent event(client_events::kShinyViewer, dataJson);
+   module_context::enqueClientEvent(event);
+}
+
+void ShinyAsyncJob::onStdout(const std::string& output)
+{
+   size_t pos = output.find(kShinyAppStarted);
+   if (pos != std::string::npos)
+   {
+      // extract the URL, which is the portion of the output string following the start token
+      url_ = output.substr(pos);
+
+      // create an event to let the client know to start viewing the running application
+      enqueueStateEvent("started");
+
+      // set the job state so the Jobs tab will show the app 
+      jobs::setJobStatus(job_, "Running");
+      setJobState(job_, jobs::JobState::JobRunning);
+
+      // no need to echo this to the user
+      return;
+   }
+
+   // forward output to base class so it can be emitted to the client
+   AsyncRJob::onStdout(output);
+}
+
+void ShinyAsyncJob::onCompleted(int exitStatus)
+{
+   if (exitStatus == 0)
+   {
+      setJobState(job_, jobs::JobState::JobSucceeded);
+      onStdout("\nShiny application finished running.\n\n");
+   }
+   else
+   {
+      if (cancelled_)
+      {
+         // typically the only way Shiny applications exit is by being stopped, so don't treat that
+         // as a failure
+         setJobState(job_, jobs::JobState::JobSucceeded);
```

Comment:
This is probably related to my above comment, but why not mark as cancelled?

Reply:
Aesthetics, mostly. Typically a Shiny process runs until you abort it. That means that as you run-edit-rerun an app, you will wind up with a big pile of aborted jobs. Having a big red `Cancelled` on every one of those jobs doesn't look nice (even though they technically were, indeed, cancelled) so we give them a nicer `Succeeded` treatment here.